src/nni_manager/training_service/remote_machine/remoteMachineTrainingService.ts

```diff
@@ -115,6 +118,50 @@ class RemoteMachineTrainingService implements TrainingService {
         }
         this.log.info('Remote machine training service exit.');
     }
+    
+    /**
+     * give trial a ssh connection
+     * @param trial 
+     */
+    public async scheduleTrialSSHClient(trial: RemoteMachineTrialJobDetail): Promise<void> {
+        const deferred: Deferred<void> = new Deferred<void>();
+        if(!trial.rmMeta) {
+            throw new Error(`rmMeta not set in trial ${trial.id}`);
+        }
+        let sshClientManager: SSHClientManager | undefined = this.machineSSHClientMap.get(trial.rmMeta);
+        if(!sshClientManager) {
+            throw new Error(`remoteSSHClient not initialized`);
+        }
+        let sshClient: Client | undefined = sshClientManager.getAvailableSSHClient();
+        if(!sshClient) {
+            //create a new connection
+            sshClient = await sshClientManager.initNewSSHClient();
+        }
+        this.trialSSHClientMap.set(trial.id, sshClient);
+
+        deferred.resolve();
+        return deferred.promise;
+    }
+    
+    /**
+     * If a trial is finished, release the connection resource
+     * @param trial 
+     */
+    public releaseTrialSSHClient(trial: RemoteMachineTrialJobDetail): void {
+        if(!trial.rmMeta) {
+            throw new Error(`rmMeta not set in trial ${trial.id}`);
+        }
+        let sshClientManager: SSHClientManager | undefined = this.machineSSHClientMap.get(trial.rmMeta);
+        if(!sshClientManager) {
+            throw new Error(`sshClientManager not initialized`);
+        }
+        const conn: Client | undefined = this.trialSSHClientMap.get(trial.id);
+        if(!conn) {
+            throw new Error(`ssh client not set in trial ${trial.id}`);
+        }
+        sshClientManager.minusUsedConnectionNumber(conn);
```

Comment:
user connection number is your implementation detail of sshClientManager, the caller doesn't need to know this. I suggest to rename minusUsedConnectionNumber to releaseConnection.

Reply:
changed to `releaseConnection`